### PR TITLE
Enhancement/update auth

### LINKED ISSUE #673: Update Authentication
**What we have**
We've updated our authentication on https://github.com/cambiatus/backend/pull/193 done on our backend

**Proposal**
Implement the necessary protocols to use the new auth

**Why**
Our security protocols on our API is pretty loose, with this we will have a regular, protected API usage on the whole stack

**How**
It will be necessary to implement the steps of our new Auth:

On `http://localhost:4000/api/graphiql`, call the gen_auth mutation:
```
mutation{
  genAuth(account:"your_account") {
    phrase
  }
}
```
Get the `phrase`
Sign with your `private_key`
Then on `http://localhost:4000/api/graphiql`, call the sign_in mutation:
```
mutation{
  signIn(
    account:"your_account",
    password:"sign_phrase"
  )
  {
    token
  }
}
```
Get the `token`

### PR DESCRIPTION
**Pull request by NeoVier**
## What issue does this PR close
Closes #673 

## Changes Proposed ( a list of new changes introduced by this PR)
- [x] Use the new auth flow described on the issue
- [x] Remove GRAPHQL_SECRET from the repo
- [x] Automatically check for errors on authenticated queries/mutations

## How to test ( a list of instructions on how to test this PR)
Check if you can login on the app and everything loads correctly! (Since this involves logging in to other communities, the latest version has been deployed to staging, so you can access communities like https://buss.staging.cambiatus.io/ to test more easily and closer to production)

There are a few edge cases to look out for:
- Try being logged in and joining a community via invite
- Try being logged out and joining a community via invite
- Try being logged in and joining a community that doesn't need invitation (through the `join` page)
- Try being logged out and joining a community that doesn't need invitation
- Replicate a user who is already logged in before this PR, and check if you remain logged in after it. In order to replicate this scenario:
  - Log into the app
  - Right click on the page > Inspect (or Inspect element in some browsers). That'll open up the dev tools
  - Find where cookies are shown (on Chrome it's Application on the top bar > Storage > Cookies > find the one with the URL you're currently on)
  - Click on the `Value` column on the `bespiral.auth_token.staging` cookie. Edit it so it's different (delete some characters, or add some characters, it really only needs to be different) - that'll make it invalid (Note: if you're on latest Chrome, you might need to [mess with some flags](https://stackoverflow.com/a/71038151) for your edited cookie to be saved).
  - Refresh the page - the app will detect the token is invalid, ask your pin, generate a new token, and rerun the query/mutation it was trying to run

### CODE REVIEW

**Comment by lucca65** [comment]
We should also remove the GRAPHQL_SECRET env var from the repo, and also from the servers

**Comment by lucca65** [comment]
One thing we should test out with this PR is the scoping of data. 

With this update users should no longer be able to query data they don't own. Here are a initial list:

- [ ] Notifications from another user (`notificationHistory`)
- [ ] Add photos to the community (`addCommunityPhotos`)
- [ ] Complete objective (`completeObjective`) 
- [ ] Delete news if not admin (`deleteNews`)
- [ ] Enable news (`hasNews`) -- what a horrible name
- [ ] Highlight News (`highlightedNews`) -- also a bad name
- [ ] create new News (`news`)

**Comment by NeoVier** [comment]
- [x] Notifications from another user (`notificationHistory`) - user is actually known by auth token, not by input on the query
- [x] Add photos to the community (`addCommunityPhotos`) - `unauthorized` error
- [x] Complete objective (`completeObjective`) - `unauthorized` error
- [x] Delete news if not admin (`deleteNews`) - `Logged user can't do that action` error -- argument should probably just be `id` instead of `newsId` -- docs only say `[Auth required]` (doesn't say it's `Admin only`)
- [x] Enable news (`hasNews`) - `unauthorized` error -- what a horrible name
- [x] Highlight News (`highlightedNews`) - `unauthorized` error -- also a bad name
- [x] create new News (`news`) - `is not admin` error

@lucca65 seems like permissions are working 💪 

I also tested them as an authorized user (the community admin), and everything worked fine, except for `deleteNews`, which produced an [error](https://cambiatus.slack.com/archives/CA83HJAAD/p1644612182918299)

**Comment by lucca65** [comment]
oh well, thanks for testing it out @NeoVier! I'll checkout the error! 💪

File: src/elm/Environment.elm [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            addStaging [ subdomain ]
+
+        subdomain :: "localhost" :: _ ->
+            [ subdomain, "staging" ]
+
+        subdomain :: "cambiatus" :: _ ->
+            [ subdomain ]
+
+        subdomain :: env :: _ ->
+            [ subdomain, env ]
```
Comment (lucca65) [inline src/elm/Environment.elm:89]: nice!

**Review by lucca65 — DISMISSED** [review]
Nice one @NeoVier!

File: src/elm/Api/Graphql.elm [hunk 2]
```diff
@@ -1,30 +1,319 @@
...
+# Error helpers
+
+@docs errorToString, isNonExistingCommunityError, isNewsNotFoundError, isAuthError
+
+---
+
+If you want to know how to get the user's private key, or how to authenticate them
+to use our EOS API, check out the `Auth` module.
+
+-}
```
[the remark this answers is not in the document]
Comment (NeoVier) [reply src/elm/Api/Graphql.elm:58]: Since our whole auth process is different from the usual, it's important to have good docs about it! 💪

**Comment by juramos-2020** [comment]
OI @NeoVier ja vou adiantando alguns feedbacks relacionados a esse teste.

**Seguem algumas obervações do teste utilizando o navegador do google chrome, versão Desktop:**

- A notificação de "NEWS" está aprecendo na tela de Welcome (aquela tela inicial que clicamos em "entrar na comunidade"). Essa notificação não precisa aparecer nessa tela.

- Ao clicar em "entrar na comunidade" está solicitando o PIN. Não me recordo se anteriormente o Pin era solicitado nessa etapa, me lembro de entrar nas comunidades sem inseri-lo, ele aparecia somente depois das 12 palavras. 

- Depois de inserir o PIN consegui entrar direto na comunidade.

**Teste utilizando a aba anônima:**

- Cliquei em "entrar na comunidade" e já fui direcionada para a tela de cadastro KYC. Não foi solicitado PIN

- Depois de inserir todas as informações para o cadastro KYC pessoa física, o cadastro não completa, aparecendo a notificação de erro "Opa. Algo de errado aconteceu enquanto criavamos sua conta"

**Comment by NeoVier** [comment]
> A notificação de "NEWS" está aprecendo na tela de Welcome (aquela tela inicial que clicamos em "entrar na comunidade"). Essa notificação não precisa aparecer nessa tela.

Feito!

> Depois de inserir todas as informações para o cadastro KYC pessoa física, o cadastro não completa, aparecendo a notificação de erro "Opa. Algo de errado aconteceu enquanto criavamos sua conta"

[Estamos investigando](https://cambiatus.slack.com/archives/CA83HJAAD/p1645216375311919)

**Comment by juramos-2020** [comment]
**Tente ser desconectado e ingressar em uma comunidade por meio de convite**

neste cenário, na página que aparece a mensagem: "Desculpe esta página é restrita
Esta comunidade requer mais algumas informações de seus membros. Para ter acesso completo, por favor, preencha os campos KYC em seu perfil", o **Botão "preencha seus dados KYC"** precisa aumentar para o texto não ficar quebrado.

Ao clicar em "sim participar" sou direcionada para tela de cadastro porém o cadastro não finaliza, aparecendo a seguinte mensagem: "Opa. Algo de errado aconteceu enquanto criavamos sua conta"


**Tente fazer login e ingressar em uma comunidade por meio de convite**

Ao tentar entrar na comunidade, com o convite e estando logado, da o seguinte erro: "Falha na autenticação"

**Tente estar logado e ingressar em uma comunidade que não precisa de convite (através da joinpágina)** e **Tente ser desconectado e ingressar em uma comunidade que não precisa de convite**

Ao clicar em "sim participar" sou direcionada para tela de cadastro porém o cadastro não finaliza, aparecendo a seguinte mensagem: "Opa. Algo de errado aconteceu enquanto criavamos sua conta"

@NeoVier , por favor, você poderia me explicar melhor o que seria isso: "Replicar um usuário que já está conectado antes deste PR" não entendi como fazer esse teste.

**Comment by NeoVier** [comment]
>  você poderia me explicar melhor o que seria isso: "Replicar um usuário que já está conectado antes deste PR" não entendi como fazer esse teste.

Claro! O objetivo do teste é ter certeza que usuários que estiverem logados antes da atualização continuem logados após a atualização.

Para o usuário continuar logado mesmo depois de recarregar a página, guardamos o seu token de autenticação no browser. Depois dessa atualização, esse token antigo não será mais válido. Então, para testar esse cenário, basta termos um token inválido guardado (ou não ter um token), e recarregar a página.

Para fazer isso (pode mudar de browser para browser, esses passos funcionam no Edge):
- Entre em alguma comunidade
- Clique com o botão direito na página > Inspect
- Ache a aba Application:
Untitled
- Nesta aba, no painel do lado esquerdo, procure por "Cookies", e clique no domínio da comunidade que você está logada:
image

- Selecione o cookie com nome `bespiral.auth_token.staging`, e o delete:

Untitled 2

- Agora é só recarregar a página. A página deve pedir o seu PIN, e então tudo deve funcionar normalmente

**Comment by juramos-2020** [comment]
OI @NeoVier 

Segue minha tela. Não estou identificando a opção ` bespiral.auth_token.staging`

image

**Comment by NeoVier** [comment]
Ah sim, nesse ambiente de testes usamos o local storage 🤦 .

As instruções são as mesmas, só que ao invés de ser na aba Cookies, é na aba Local Storage > https://deploy-preview...., e o nome do item é `bespiral.auth_token.netlify`

**Comment by NeoVier** [comment]
Boa @lucca65 🔥 🚀 

@juramos-2020 todos os problemas relacionados a comunidades com KYC devem estar resolvidos 🤞. Aproveitei pra trocar a ordem das coisas: ao invés de primeiro preencher as informações de KYC, e depois clicar em se juntar à comunidade, acho melhor primeiro o usuário entrar na comunidade, para então preencher as informações de KYC.

---
> Ao tentar entrar na comunidade, com o convite e estando logado, da o seguinte erro: "Falha na autenticação"

Não consegui replicar esse erro, pra mim funcionou normal entrar em uma comunidade com convite. Pode ver se consegue replicar de novo, por favor?

**Comment by juramos-2020** [comment]
`Agora é só recarregar a página. A página deve pedir o seu PIN, e então tudo deve funcionar normalmente`
ok tudo funcionou como o esperado 👍 


**Tente fazer login e ingressar em uma comunidade por meio de convite**
ok tudo funcionou como o esperado 👍 

Tente ser desconectado e ingressar em uma comunidade por meio de convite
ok tudo funcionou como o esperado 👍 

Tente estar logado e ingressar em uma comunidade que não precisa de convite (através da joinpágina)
ok tudo funcionou como o esperado 👍 

🔥 🚀🔥 🚀🔥 🚀

Tente ser desconectado e ingressar em uma comunidade que não precisa de convite
ok tudo funcionou como o esperado 👍 

Também testei o cenário de preencher o formulário KYC na aba anônima e dessa vez deu certo  👍

**Comment by lucca65** [comment]
UHUL!
